### PR TITLE
MGMT-10597: CI for ZTP Day2 SNO

### PR DESCRIPTION
**Pull request by paul-maidment**
This commit adds a test of the ZTP functionality for adding a Day1
single-node cluster followed by the addition of a worker node to this
cluster as a Day2 ZTP operation.

### CODE REVIEW

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:63]: missing `|`

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-commands.sh
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-commands.sh:69]: Missing `|`

File: ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml
Comment (omertuc) [inline ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml:377]: Name should reflect it's SNO + workers

Comment (omertuc) [inline ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml:377]: We also need to add a periodic job, not just an assisted-service presubmit

Comment (omertuc) [reply ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml:377]: Make sure the periodic job also has a `reporter_config`

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:19]: Please add a more verbose documentation, explaining that when the number of workers is 0, this step does nothing

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:64]: Remove unnecessary line

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:69]: Add a comment above this line that explains that we only use the first (n - NUMBER_OF_DAY2_HOSTS) baremetal hosts for the spoke cluster, and the rest of them are used for day-2 by the "maybe" step

File: ci-operator/jobs/openshift/assisted-service/openshift-assisted-service-master-presubmits.yaml
Comment (osherdp) [inline ci-operator/jobs/openshift/assisted-service/openshift-assisted-service-master-presubmits.yaml:1836]: this means it will run for each PR

**Review by osherdp — CHANGES_REQUESTED** [review]
/hold
Partial review, I'll take another look later on

**Comment by osherdp** [comment]
> Once it's merged will that automatically added to a periodic CI or do we need to do something for it?

No, this PR only adds it as a presubmit job to assisted-service repo

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml
Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: The maybe clarifies that it's not weird to use this step in the regular (non-day2) workflow

**Comment by omertuc** [comment]
> > Once it's merged will that automatically added to a periodic CI or do we need to do something for it?
> 
> No, this PR only adds it as a presubmit job to assisted-service repo

See https://github.com/openshift/release/pull/28958#discussion_r902406967

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml
Comment (paul-maidment) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: So rename to `assisted-baremetal-operator-ztp-add-day2-workers` ?

Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: I disagree, I think `maybe` is important

Comment (osherdp) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: so at least let's make it ``assisted-baremetal-operator-ztp-optionally-add-day2-worker``
I've never seen "maybe" in any piece of code

Comment (osherdp) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: Just one optionally... ``assisted-baremetal-operator-ztp-add-day2-workers-optionally`` :smiley:

Comment (paul-maidment) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: I'm going with `assisted-baremetal-operator-ztp-add-day2-workers-optionally`

Comment (omertuc) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/maybe-add-day2-worker/assisted-baremetal-operator-ztp-maybe-add-day2-worker-ref.yaml:4]: >Just one "optionally" is needed... assisted-baremetal-operator-ztp-add-day2-workers-optionally smiley

Oops, I hope you got what I meant.

Anyway, great bike shedding session thank you for coming

File: ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml
Comment (omertuc) [inline ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml:390]: We only need `100`

File: ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml
Comment (omertuc) [reply ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml]: Do not change existing jobs.

Most jobs relating to single-node across the OpenShift organization use `single-node`

Comment (paul-maidment) [reply ci-operator/config/openshift/assisted-service/openshift-assisted-service-master__edge.yaml]: Actually found another problem, if we use `single-node` instead of `sno` it breaks a validation on the length of the job name. I'm open to suggestions about full job name that would include `single-node`? while not breaking this?

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh:66]:
```suggestion
# These are the last NUMBER_OF_DAY2_HOSTS in this file: '/root/dev-scripts/${EXTRA_BAREMETALHOSTS_FILE}'
```

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml
Comment (omertuc) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml:19]: Please document NUMBER_OF_DAY2_HOSTS here and explain that it's optional and doesn't add any workers when that variable is set to 0

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (paul-maidment) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:70]: Done

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (paul-maidment) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:65]: Done

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh:7]: nit: update this with the step name

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh:27]: we're already doing all of this in the previous step, no?

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml:12]: from what I'm seeing, the step doesn't use this env field

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:32]: thanks for handling that!

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:57]:
```suggestion
# Inject job configuration for ZTP, if available
```

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh:70]:
```suggestion
  echo "No day 2 hosts defined, skipping day 2 hosts"
  exit 0
<no-else>
```

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml:15]: we can possibly handle without this field as well

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh
Comment (osherdp) [inline ci-operator/step-registry/assisted/baremetal/operator/ztp/assisted-baremetal-operator-ztp-commands.sh:68]: fix indentation to be like the rest of the file

**Review by osherdp — CHANGES_REQUESTED** [review]

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml
Comment (paul-maidment) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-ref.yaml:22]: See lines above that document an env var...

File: ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh
Comment (paul-maidment) [reply ci-operator/step-registry/assisted/baremetal/operator/ztp/add-day2-workers-optionally/assisted-baremetal-operator-ztp-add-day2-workers-optionally-commands.sh:52]: I made this change now

**Comment by osherdp** [comment]
I think non of those failures are related to your changes, so I'm triggering some more rehearses
/test pj-rehearse-more

**Comment by omertuc** [comment]
Waiting for osher to lgtm

**Comment by omertuc** [comment]
/lgtm

@osherdp unhold if you approve

**Review by osherdp — APPROVED** [review]
Looking good!
/unhold

**Pull request merged**
